libr/io/map.c

```diff
@@ -1,12 +1,132 @@
-/* radare2 - LGPL - Copyright 2017 - condret */
+/* radare2 - LGPL - Copyright 2017 - condret, MaskRay */
 
-#include <r_io.h>
-#include <r_util.h>
-#include <sdb.h>
 #include <stdlib.h>
+#include <sdb.h>
+#include "r_binheap.h"
+#include "r_io.h"
+#include "r_util.h"
+#include "r_vector.h"
 
 #define END_OF_MAP_IDS  0xffffffff
 
+#define MAP_USE_HALF_CLOSED 0
+
+struct map_event_t {
+	RIOMap *map;
+	ut64 addr;
+	int id; // distinct priority in [0, len(maps))
+	bool is_to;
+};
+
+// Sort by address, (addr, !is_to) precedes (addr, is_to)
+static int _cmp_map_event(const void *a_, const void *b_) {
+	struct map_event_t *a = (void *)a_, *b = (void *)b_;
+	if (a->addr != b->addr) {
+		return a->addr < b->addr ? -1 : 1;
+	}
+	return a->is_to - b->is_to; // TODO swap if half-closed
+}
+
+static int _cmp_map_event_by_id(const void *a_, const void *b_) {
+	struct map_event_t *a = (void *)a_, *b = (void *)b_;
+	return a->id - b->id;
+}
+
+static bool _map_skyline_push(RVector *map_skyline, ut64 from, ut64 to, RIOMap *map) {
+	RIOMapSkyline *part = R_NEW (RIOMapSkyline);
+	if (!part) {
+		return false;
+	}
+	part->map = map;
+	part->from = from;
+	part->to = to - 1; // TODO half-closed
```

Comment:
yeah im all about removing those ugly -1's

Reply:
can you pls wait a bit, I'm ok with removing closed intervals. Even if i don't like the half opened intervals, because their mathematical modell breaks at UT64_MAX, I'm ok with that change. But please wait a bit. There are things that would potentially get broken by that change, and I want to add tests befor